File: src/main/webapp/assets/js/space/spaceship/Bullet.js
Engine.define("Bullet", ['Vector', 'Geometry', 'BaseObject', 'Explosion', 'Point'], function() {

    var BaseObject = Engine.require("BaseObject");
    var Explosion = Engine.require("Explosion");
    var Geometry = Engine.require("Geometry");
    var Vector = Engine.require("Vector");
    var Point = Engine.require("Point");

    function Bullet(x, y, angle) {
        BaseObject.apply(this, arguments);
        this.vector = null;
        var me = this;
        this.ship = null;
        this.alive = true;
        setTimeout(function(){
            me.alive = false;
        }, 3000);
    }
    Bullet.prototype = Object.create(BaseObject.prototype);
    Bullet.constructor = Bullet;


    Bullet.prototype.correct = function(spaceShip) {
        //this.points = [new Point(oldX, ol
        this.x += spaceShip.x;
        this.y += spaceShip.y;
        var translated = Geometry.translate(spaceShip, this, spaceShip.angle);
        this.x = translated.x;
        this.y = translated.y;
        this.angle += spaceShip.angle;

        this.points = [
            new Point(this.x, this.y),
            new Point(this.x + 5 * Math.cos(this.angle), this.y + 5 * Math.sin(this.angle))
        ];
        this.oldPoint = this.points[0];
        this.vector = new Vector(
            Bullet.speed * Math.cos(this.angle),
            Bullet.speed * Math.sin(this.angle)
        );
        this.vector.append(spaceShip.vector);
        this.ship = spaceShip.id;
    };

    Bullet.speed = 5;

    Bullet.prototype.draw = function(context, appContext) {
        var shift = appContext.shift;
        context.save();
        context.translate(this.x + shift.x, this.y+ shift.y);
        context.rotate(this.angle);
        context.beginPath();
        context.strokeStyle = "white";

        context.moveTo(-3, 0);
        context.lineTo(2, 0);
        context.stroke();
        context.restore();

        context.beginPath();
        context.strokeStyle = "red";
        context.moveTo(this.points[0].x+ shift.x, this.points[0].y+ shift.y);
        context.lineTo(this.points[1].x+ shift.x, this.points[1].y+ shift.y);
        context.stroke();



    };
    Bullet.prototype.update = function() {
        this.x += this.vector.x;
        this.y += this.vector.y;
        this.oldPoint = this.points[0].clone();
        this.points[0].x += this.vector.x;
        this.points[0].y += this.vector.y;
        this.points[1].x += this.vector.x;
        this.points[1].y += this.vector.y;
    };

    Bullet.prototype.isAlive = function() {
        return this.alive;
    };

    Bullet.prototype.hasImpact = function(object) {
        if(object.id !== this.ship) {
            return BaseObject.prototype.hasImpact.apply(this, arguments);
        } else {
            return false
        }
    };

    Bullet.prototype.onImpact = function(object) {
        this.alive = false;
        return [new Explosion(this.x, this.y, this.vector, 8, 60)];
    };

    Bullet.prototype.getPoints = function(){
        if(!this.oldPoint || !this.points[1]) {
            return [];
        } else {
            return [this.oldPoint, this.points[1]];
        }
    };

    return Bullet;
});

File: src/main/java/org/forweb/drift/entity/drift/Bullet.java
package org.forweb.drift.entity.drift;


import com.fasterxml.jackson.annotation.JsonIgnore;
import org.forweb.drift.entity.drift.spaceships.SpaceShip;
import org.forweb.drift.utils.IncrementalId;
import org.forweb.geometry.misc.Angle;
import org.forweb.geometry.misc.Vector;
import org.forweb.geometry.services.PointService;
import org.forweb.geometry.shapes.Point;
import org.jbox2d.common.Vec2;

public class Bullet extends BaseObject {

    static private double getSpeed() {
        return 5;
    }

    private int ship;
    private Vec2 oldPoint;
    private double creationTime;

    public Bullet(double x, double y, double angle, int id) {
        super(x, y, angle, id);
        this.creationTime = System.currentTimeMillis() + 3000;
    }

    @Override
    public String getType() {
        return Types.bullet.toString();
    }

    @JsonIgnore
    @Override
    public boolean isRelativePoints() {
        return false;
    }

    public boolean isAlive() {
        return !(this.creationTime == 0) && !(System.currentTimeMillis() > this.creationTime);
    }

    public void correct(SpaceShip spaceShip) {
        setX(getX() + spaceShip.getX());
        setY(getY() + spaceShip.getY());
        Point translated = PointService.translate(
                new Point(spaceShip.getX(), spaceShip.getY()),
                new Point(getX(), getY()),
                spaceShip.getAngle()
        );
        setX(translated.getX());
        setY(translated.getY());
        Angle angle = getAngle();
        angle.append(spaceShip.getAngle());

        Vec2[] points = new Vec2[2];
        points[0] =new Vec2(this.getX(), this.getY());
        oldPoint = points[0];
        points[1] = new Vec2(this.getX() + 5 * angle.cos(), this.getY() + 5 * angle.sin());
        this.setPoints(points);

        setVector(
                new Vector(
                        Bullet.getSpeed() * angle.cos(),
                        Bullet.getSpeed() * angle.sin()
                )
        );
        Vector vector = getVector();
        vector.append(spaceShip.getVector());
        this.ship = spaceShip.getId();
    }

    public BaseObject[] update() {
        this.oldPoint = this.points[0];
        Vector vector = getVector();
        this.setX(getX() + vector.x);
        this.setY(getY() + vector.y);
        this.points[0] = new Vec2(this.points[0].x + vector.x, this.points[0].y + vector.y);
        this.points[1] = new Vec2(this.points[1].x + vector.x, this.points[1].y + vector.y);
        return null;
    }

    @Override
    public Vec2[] getPoints() {
        Vec2[] out = new Vec2[2];
        out[0] = oldPoint;
        out[1] = this.points[1];
        return out;
    };

    @Override
    public BaseObject[] onImpact(BaseObject object, IncrementalId ids) {
        BaseObject[] out = new BaseObject[1];
        out[0] = new Explosion(getX(), getY(), getVector(), 8, 60, ids.get());
        this.creationTime = 0;
        return out;
    }

    public boolean hasImpact(BaseObject baseObject) {
        if(baseObject instanceof SpaceShip && baseObject.getId() == this.ship) {
            return false;
        } else {
            return super.hasImpact(baseObject);
        }
    }

    public int getShip() {
        return ship;
    }
}